packages/coverage-c8/src/provider.ts

```diff
@@ -50,37 +50,44 @@ export class C8CoverageProvider implements CoverageProvider {
 
     // add source maps
     const sourceMapMeta: Record<string, { map: RawSourceMap; source: string | undefined }> = {}
-    await Promise.all(Array
+
+    const entries = Array
```

Comment:
If `Promise.all` is preferred for speed, I guess another solution would be to prepare this `entries` list so that it does not contain these "duplicate" entries. Going through list of unique files in `Promise.all` should be ok.

Reply:
I would prefer `Promise.all`. The only things that's asynchronies is reading a file. We can do this safely, I think.

Reply:
Or maybe we should also merge source maps?

Reply:
Refactored the code so that `entries` is a list of unique files, meaning that `file.vue` and `file.vue?query...` are considered the same. Using `Promise.all` is now safe since separate Promises won't be overwriting each others' results anymore.

But still, it is now counting on the fact that `file.vue` will be loaded before `file.vue?query...`. The latter one is the case where soucemaps break. Can be rely on this order? 

> Or maybe we should also merge source maps?

How would this work? Any examples or pointers?

Reply:
> But still, it is now counting on the fact that file.vue will be loaded before file.vue?query....

We can for Vue files, but we can't overall, I think. It depends on the compiler.

> How would this work?

I don't actually know what's inside source maps right now, so maybe it won't work, but Vite for example has this function: https://github.com/vitejs/vite/blob/47a78db49740daee873525de90d0e44b19883771/packages/vite/src/node/utils.ts#L735

Reply:
I don't think there is much to merge. The first file request looks OK, but the second one is completely empty.

<details>
  <summary>First request, without query params</summary>

```json
{
  "filepath": "/Users/x/y/demo-c8-issues/src/components/StyleTag.vue",
  "url": "file:///Users/x/y/demo-c8-issues/src/components/StyleTag.vue",
  "map": {
    "version": 3,
    "mappings": ";;;;;;AACA,QAAI,OAAO;AACT,cAAQ,IAAI,OAAO;AAAA,IACrB,OAAO;AACL,cAAQ,IAAI,MAAM;AAAA,IACpB;;;;;;;SAGA",
    "names": [],
    "sources": [
      "/Users/x/y/demo-c8-issues/src/components/StyleTag.vue"
    ],
    "sourcesContent": [
      "<script setup lang=\"ts\">\nif (false) {\n  console.log('false');\n} else {\n  console.log('true');\n}\n</script>\n\n<template>Component with style tag</template>\n\n<style lang=\"css\" scoped>\nbody {\n  background-color: #bada55;\n}\n</style>\n"
    ],
    "file": "/Users/x/y/demo-c8-issues/src/components/StyleTag.vue"
  },
  "source": "const __vite_ssr_import_0__ = await __vite_ssr_import__(\"/node_modules/.pnpm/vue@3.2.45/node_modules/vue/index.mjs\");\n\nconst _sfc_main = /* @__PURE__ */ __vite_ssr_import_0__.defineComponent({\n  __name: \"StyleTag\",\n  setup(__props, { expose }) {\n    expose();\n    if (false) {\n      console.log(\"false\");\n    } else {\n      console.log(\"true\");\n    }\n    const __returned__ = {};\n    Object.defineProperty(__returned__, \"__isScriptSetup\", { enumerable: false, value: true });\n    return __returned__;\n  }\n});\nfunction _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {\n  return \"Component with style tag\";\n}\nconst __vite_ssr_import_1__ = await __vite_ssr_import__(\"/src/components/StyleTag.vue?vue&type=style&index=0&scoped=88d9cf36&lang.css\");\n\nconst __vite_ssr_import_2__ = await __vite_ssr_import__(\"/@id/__x00__plugin-vue:export-helper\");\n\n__vite_ssr_exports__.default = /* @__PURE__ */ __vite_ssr_import_2__.default(_sfc_main, [[\"render\", _sfc_render], [\"__scopeId\", \"data-v-88d9cf36\"], [\"__file\", \"/Users/x/y/demo-c8-issues/src/components/StyleTag.vue\"]]);\n\n//# sourceMappingSource=vite-node\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJtYXBwaW5ncyI6Ijs7Ozs7O0FBQ0EsUUFBSSxPQUFPO0FBQ1QsY0FBUSxJQUFJLE9BQU87QUFBQSxJQUNyQixPQUFPO0FBQ0wsY0FBUSxJQUFJLE1BQU07QUFBQSxJQUNwQjs7Ozs7OztTQUdBIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbIi9Vc2Vycy9hcmkvR2l0L2RlbW8tYzgtaXNzdWVzL3NyYy9jb21wb25lbnRzL1N0eWxlVGFnLnZ1ZSJdLCJzb3VyY2VzQ29udGVudCI6WyI8c2NyaXB0IHNldHVwIGxhbmc9XCJ0c1wiPlxuaWYgKGZhbHNlKSB7XG4gIGNvbnNvbGUubG9nKCdmYWxzZScpO1xufSBlbHNlIHtcbiAgY29uc29sZS5sb2coJ3RydWUnKTtcbn1cbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+Q29tcG9uZW50IHdpdGggc3R5bGUgdGFnPC90ZW1wbGF0ZT5cblxuPHN0eWxlIGxhbmc9XCJjc3NcIiBzY29wZWQ+XG5ib2R5IHtcbiAgYmFja2dyb3VuZC1jb2xvcjogI2JhZGE1NTtcbn1cbjwvc3R5bGU+XG4iXSwiZmlsZSI6Ii9Vc2Vycy9hcmkvR2l0L2RlbW8tYzgtaXNzdWVzL3NyYy9jb21wb25lbnRzL1N0eWxlVGFnLnZ1ZSJ9\n"
}
```

</details>

<details>
  <summary>Second request, with query params</summary>

```json
{
  "filepath": "/Users/x/y/demo-c8-issues/src/components/StyleTag.vue",
  "url": "file:///Users/x/y/demo-c8-issues/src/components/StyleTag.vue",
  "map": {
    "version": 3,
    "file": null,
    "sources": [
      "/Users/x/y/demo-c8-issues/src/components/StyleTag.vue?vue&type=style&index=0&scoped=88d9cf36&lang.css"
    ],
    "sourcesContent": [
      ""
    ],
    "names": [],
    "mappings": ""
  },
  "source": "\n\n//# sourceMappingSource=vite-node\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjpudWxsLCJzb3VyY2VzIjpbIi9Vc2Vycy9hcmkvR2l0L2RlbW8tYzgtaXNzdWVzL3NyYy9jb21wb25lbnRzL1N0eWxlVGFnLnZ1ZT92dWUmdHlwZT1zdHlsZSZpbmRleD0wJnNjb3BlZD04OGQ5Y2YzNiZsYW5nLmNzcyJdLCJzb3VyY2VzQ29udGVudCI6WyIiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IiJ9\n"
}
```

</details>

Should we instead compare these two maps when duplicate entries are encountered, and do some sanity checks to the existing entry to see whether it should be overwritten? e.g. `mappings.length > 0`, `sourcesContent[0].length > 0`?

Reply:
Sure, let's add some sanity check here.

Reply:
Alright, there is now following filter:

```ts
.filter((entry) => {
  if (!entry)
    return false

  // Mappings and sourceContent are needed for C8 to work
  return (
    entry.map.mappings.length > 0
    && entry.map.sourcesContent
    && entry.map.sourcesContent.length > 0
    && entry.map.sourcesContent[0].length > 0
  )
}) 
```

This will simply drop out the sourcemaps that cannot be used and are considered as broken. It doesn't specifically look for duplicate entries, but prevents breaking the coverage. Any thoughts?

Reply:
Seems fine to me, Do we need to add a test for this case?

Reply:
Now using `test/coverage-test/src/Defined.vue` for testing this case. In case of regression, C8 provider's JSON report test will report lots of uncovered/missing lines. As this is happened randomly, tests will become flaky if implementation breaks.

Reply:
> As this happened randomly, tests will become flaky if implementation breaks.

To be honest, they are quite flaky already 👀 I noticed the coverage file doesn't exist sometimes (~20%)

Reply:
Oh wow, I don't think I've run into that issue at all. At least locally tests are passing 100% constantly on my machine.

Reply:
Example: https://github.com/vitest-dev/vitest/actions/runs/3841585602/jobs/6541978570

Reply:
The test assertions pass, but clean up fails. That's weird.

> Error: ENOTEMPTY: directory not empty, rmdir '/home/runner/work/vitest/vitest/test/coverage-test/coverage'

Coming from here, even though it is using the `force` option:

https://github.com/vitest-dev/vitest/blob/199776767a10bcd9f241e36a5cdd578c7d7c0dc2/packages/coverage-c8/src/provider.ts#L35-L37


I'll debug this further and see if I can reproduce it.

<details>
  <summary>Relevant logs from CI failure</summary>

```
test/coverage-test test: > @vitest/test-coverage@ test:c8 /home/runner/work/vitest/vitest/test/coverage-test
test/coverage-test test: > node ./testing.mjs --provider c8

test/coverage-test test:  RUN  v0.26.3 /home/runner/work/vitest/vitest/test/coverage-test
test/coverage-test test:       Coverage enabled with c8
test/coverage-test test:  ✓ test/vue.test.ts > vue 3 coverage
test/coverage-test test:  ✓ test/vue.test.ts > define package in vm
test/coverage-test test:  ✓ test/vue.test.ts > vue non-SFC, uses query parameters in file imports
test/coverage-test test:  ✓ test/coverage.test.ts > Math.sqrt()
test/coverage-test test:  ✓ test/coverage.test.ts > Math.sqrt()
test/coverage-test test:  ✓ test/coverage.test.ts > implicit else
test/coverage-test test:  ✓ test/coverage.test.ts > import meta env
test/coverage-test test:  Test Files  2 passed (2)
test/coverage-test test:       Tests  6 passed (6)
test/coverage-test test:    Start at  21:12:00
test/coverage-test test:    Duration  8.46s (transform 2.67s, setup 0ms, collect 1.51s, tests 210ms)
test/coverage-test test:  % Coverage report from c8
test/coverage-test test: -----------------------|---------|----------|---------|---------|-------------------
test/coverage-test test: File                   | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
test/coverage-test test: -----------------------|---------|----------|---------|---------|-------------------
test/coverage-test test: All files              |   69.17 |       90 |      64 |   69.17 |                   
test/coverage-test test:  src                   |   60.57 |    86.66 |   63.15 |   60.57 |                   
test/coverage-test test:   Defined.vue          |     100 |      100 |     100 |     100 |                   
test/coverage-test test:   Hello.vue            |     100 |      100 |     100 |     100 |                   
test/coverage-test test:   implicitElse.ts      |     100 |      100 |     100 |     100 |                   
test/coverage-test test:   importEnv.ts         |     100 |      100 |     100 |     100 |                   
test/coverage-test test:   index.mts            |     100 |      100 |     100 |     100 |                   
test/coverage-test test:   untested-file.ts     |       0 |        0 |       0 |       0 | 1-33              
test/coverage-test test:   utils.ts             |   72.41 |    85.71 |      50 |   72.41 | 10-12,22-24,28-29 
test/coverage-test test:  src/Counter           |   90.47 |      100 |   66.66 |   90.47 |                   
test/coverage-test test:   Counter.component.ts |   83.33 |      100 |      50 |   83.33 | 13-14,21-22       
test/coverage-test test:   Counter.vue          |     100 |      100 |     100 |     100 |                   
test/coverage-test test:   index.ts             |     100 |      100 |     100 |     100 |                   
test/coverage-test test: -----------------------|---------|----------|---------|---------|-------------------

test/coverage-test test:  RUN  v0.26.3 /home/runner/work/vitest/vitest/test/coverage-test
test/coverage-test test:  ✓ coverage-report-tests/c8.report.test.ts > c8 json report
test/coverage-test test:  ✓ coverage-report-tests/generic.report.test.ts > html report
test/coverage-test test:  ✓ coverage-report-tests/generic.report.test.ts > html report
test/coverage-test test:  ✓ coverage-report-tests/generic.report.test.ts > lcov report
test/coverage-test test:  ✓ coverage-report-tests/generic.report.test.ts > all includes untested files
test/coverage-test test:  ✓ coverage-report-tests/generic.report.test.ts > files should not contain query parameters
test/coverage-test test:  ✓ coverage-report-tests/generic.report.test.ts > file using import.meta.env is included in report
test/coverage-test test:  Test Files  2 passed (2)
test/coverage-test test:       Tests  6 passed (6)
test/coverage-test test:    Start at  21:12:10
test/coverage-test test:    Duration  5.12s (transform 1.21s, setup 0ms, collect 169ms, tests 241ms)
test/coverage-test test: [Error: ENOTEMPTY: directory not empty, rmdir '/home/runner/work/vitest/vitest/test/coverage-test/coverage'] {
test/coverage-test test:   errno: -39,
test/coverage-test test:   code: 'ENOTEMPTY',
test/coverage-test test:   syscall: 'rmdir',
test/coverage-test test:   path: '/home/runner/work/vitest/vitest/test/coverage-test/coverage'
test/coverage-test test: }
test/coverage-test test:  ELIFECYCLE  Command failed with exit code 1.
test/coverage-test test: Failed

```

</details>

Reply:
Btw there are similar errors coming from reporter tests: https://github.com/vitest-dev/vitest/actions/runs/3841585602/jobs/6541978793. Maybe this is not directly related to coverage tests. 

> Error: ENOTEMPTY: directory not empty, rmdir '\\?\D:\a\vitest\vitest\test\reporters\junitReportDirectory\deeply\nested'

There's also some weird logging coming from Node itself:

> test/reporters test: (node:3912) Warning: Closing file descriptor 7 on garbage collection
> test/reporters test: (Use `node --trace-warnings ...` to show where the warning was created)
> test/reporters test: (node:3912) [DEP0137] DeprecationWarning: Closing a FileHandle object on garbage collection is deprecated. Please close FileHandle objects explicitly using FileHandle.prototype.close(). In the future, an error will be thrown if a file descriptor is closed during garbage collection.
> test/reporters test: (node:3912) Warning: Closing file descriptor 6 on garbage collection
> test/reporters test: (node:3912) Warning: Closing file descriptor 3 on garbage collection
> test/reporters test: (node:3912) Warning: Closing file descriptor 5 on garbage collection
> test/reporters test: (node:3912) Warning: Closing file descriptor 4 on garbage collection

As this seems to be coming on Node 18 only, I wonder if there is some recent change in Node that causes this.

Reply:
Windows just fails every time (I am looking into it), but coverage fails sometimes: https://github.com/vitest-dev/vitest/actions/runs/3854051365/jobs/6567634858 (this is successful run for coverage). Coverage also failed on node 16 previously, so I don't this is related.

Reply:
I've tried to track this here: 
- https://github.com/AriPerkkio/vitest/actions/runs/3853805610
- https://github.com/AriPerkkio/vitest/commit/bf86ba1bf86a841f1a333907c3bd3792410c468b

It's running coverage and reporter tests on all 18.x versions, on each OS. Reporter tests seem to fail on Windows only, but coverage tests fail on Mac and Linux too. The failures seem really random, I don't see any patterns there...